countdown and message models (closes #2464)

FinnStutzenstein opened the pull request:
@normanjaeckel I am trying to create two new projection defaults in the migration file. Skipping the autoupdate works. But: this code deletes all old projectiondefaults?? If you comment out the RunPython, you can see all projectiondefaults in the rest api; with the RunPython only the new message and countdown projectiondefaults are available.

normanjaeckel commented:
The current projection default instances are created by a signal receiver. This receiver does nothing if there are some projection defaults already in the database. So it is currently not possible to add new projection defaults during migration. One solution might be to move all the code to a migration file.

FinnStutzenstein commented:
Ah, right; I was creating a new database without any projectiondefaults. I think for compatibility I move all Projectiondefaults to a seperate Migrationfile.

FinnStutzenstein commented:
@emanuelschuetze @normanjaeckel ready to test

emanuelschuetze commented:
I found a problem while testing:
* Start with fresh db.
* create new countdown and project it to projector 1.
* delete countdown (while projecting) 
-> Projector shows "NaN:aN:aN". Rest API /rest/core/projector/ still shows projector element.
* Reload site and show projector live view.
-> Gets JS error "Error for slide core/countdown: Countdown does not exists". Rest API shows also now an error. No possibility to remove this invalid projector element. In projector sidebar there is no countdown listed.
Same problem for messages!
So this is a blocker before we can merge this pull request.

FinnStutzenstein commented:
@emanuelschuetze this is a general misbehaviour in design. Try to protect something else and then delete ist (e.g. motions, topics...) This should always give this error. On delete the server should check, if the object is projected and if true, remove it from the projector elements.

emanuelschuetze commented:
There is a difference between countdowns and e.g. topic slides:
* If you delete a projected topic slide, the projector shows an empty slide.
* If you delete a countdown, the projector shows invalid time format "NaN:aN:aN" -> expected: does not show this on projector after deletion.
* In current master the rest api clears the deleted projector elements after project a new topic slide or countdown. With your branch I can't remove the invalid projector element from rest. Right?

FinnStutzenstein commented:
The problem is that countdowns (and the same for messages) are stable. If you delete a topic, you have an invalid slide, but if you project something else, all non stable elements get cleared, so the invalid slide.

But stable elements do not get cleared; so this countdown element (from the deleted countdown) will stay in the config forever.

Question of design: If you delete a projected topic an empty topic slide is shown. But I would prefer to just remove the deleted topic from the projector config. If we do this for every model (check if it is projected, and remove it from the projector config) the problem with messages and countdowns are solved. Your opinion?

emanuelschuetze commented:
> The problem is that countdowns (and the same for messages) are stable

Ah, right...

@normanjaeckel What do you think?

normanjaeckel commented:
Hm. Of course it is possible to check if there are active slides on delete of **every** django model. This is a huge load because we may have a lot of slides that require this or that model instance to exist. I would prefer a way that clears only the stable countdowns on delete.

emanuelschuetze commented:
> I would prefer a way that clears only the stable countdowns on delete.

Can we do this in an extra pull request?

normanjaeckel commented:
I propose that the client sends an additional request to the server to deactivate the countdown projector element on delete.

FinnStutzenstein commented:
@emanuelschuetze ping.

FinnStutzenstein commented:
Note: If PR #2558  is merged erlier, this needs some changes.

emanuelschuetze commented:
@FinnStutzenstein #2558 is merged. Please make your changes here.

emanuelschuetze commented:
... and rebase all together

FinnStutzenstein commented:
@emanuelschuetze change `project()` functions of messages and countdowns. Ready to merge from m point of view ;)

normanjaeckel commented on openslides/agenda/static/js/agenda/site.js:
```diff
@@ -431,6 +431,9 @@ angular.module('OpenSlidesApp.agenda.site', [
         // Return true if the requested user is allowed to do a specific action
         // and see the corresponding button (e.g. 'add me' or 'remove me').
         $scope.isAllowed = function (action) {
+            if (!operator.user) {
```
Why should anonymous not be allowed to start or stop speakers (if the default group has this permission of course)?

FinnStutzenstein replied:
See Issue #2588 . If the operator is not loaded this code would throw errors.

normanjaeckel commented on openslides/core/models.py:
```diff
@@ -294,6 +297,76 @@ def __str__(self):
+
+    message = models.TextField(blank=True)
+
+    identify = models.BooleanField(default=False)
```
I do not understand why we want to save the identify message into the database.

FinnStutzenstein replied:
It's just a boolean representing if this message is an identifier message or not.

normanjaeckel replied:
As discussed with @emanuelschuetze. Please try to remove this field.

normanjaeckel commented on openslides/core/models.py:
```diff
@@ -294,6 +297,76 @@ def __str__(self):
+
+    class Meta:
+        default_permissions = ()
+        permissions = ()
```
This line can be removed I think.

normanjaeckel commented on openslides/core/models.py:
```diff
@@ -294,6 +297,76 @@ def __str__(self):
+
+    class Meta:
+        default_permissions = ()
+        permissions = ()
```
This line can be removed.

normanjaeckel commented on openslides/core/models.py:
```diff
@@ -294,6 +297,76 @@ def __str__(self):
+        default_permissions = ()
+        permissions = ()
+
+    def control(action, index=1):
```
Missing self attribute. (I think we do not want to have a staticmethod here.)

normanjaeckel commented on openslides/core/models.py:
```diff
@@ -294,6 +297,76 @@ def __str__(self):
+            raise ValueError("Action must be 'start', 'stop' or 'reset', not {}.".format(action))
+
+        try:
+            countdown = Countdown.objects.get(pk=index)
```
If this is an instance method, countdown is always self.

normanjaeckel commented on openslides/core/models.py:
```diff
@@ -294,6 +297,76 @@ def __str__(self):
+            countdown = Countdown(
+                pk=index,
+                default_time=config['projector_default_countdown'],
+                countdown_time=60,
```
countdown_time=config['projector_default_countdown'] I think

normanjaeckel commented on openslides/core/projector.py:
```diff
@@ -15,134 +10,41 @@ class Clock(ProjectorElement):
 
 
-class Countdown(ProjectorElement):
+class CountdownSlide(ProjectorElement):
```
Please rename this to CountdownElement

normanjaeckel commented on openslides/core/projector.py:
```diff
@@ -15,134 +10,41 @@ class Clock(ProjectorElement):
 
 
-class Message(ProjectorElement):
+class MessageSlide(ProjectorElement):
```
Please rename this to ProjectorMessageElement

normanjaeckel commented on openslides/core/models.py:
```diff
@@ -294,6 +297,76 @@ def __str__(self):
         return 'Message {}'.format(self.timestamp)
 
 
+class Message(RESTModelMixin, models.Model):
```
Please rename this model to ProjectorMessage. Also rename other classes respectivly.

normanjaeckel commented on openslides/core/static/js/core/site.js:
```diff
@@ -1157,27 +984,35 @@ angular.module('OpenSlidesApp.core.site', [
+                    identify: true,
+                };
+                Message.create(message).then(function(message){
+                    console.log("created:");
```
Remove this line.

normanjaeckel commented on openslides/core/static/js/core/site.js:
```diff
@@ -1157,27 +984,35 @@ angular.module('OpenSlidesApp.core.site', [
+                };
+                Message.create(message).then(function(message){
+                    console.log("created:");
+                    console.log(message);
```
Remove this line.

FinnStutzenstein commented:
@normanjaeckel ping.

emanuelschuetze commented:
@FinnStutzenstein travis failed

FinnStutzenstein commented:
@emanuelschuetze missed one piece of code.. Now fixed.

FinnStutzenstein commented:
@emanuelschuetze Please retest.

normanjaeckel commented on openslides/core/models.py:
```diff
@@ -294,6 +297,59 @@ def __str__(self):
+        default_permissions = ()
+
+
+class ProjectorCountdown(RESTModelMixin, models.Model):
```
The model can be named Countdown. I only wanted the model Message to be renamed to ProjectorMessage because we also have ChatMessage here. But there is only one Countdown. Sorry ...

FinnStutzenstein replied:
Done.

normanjaeckel commented on openslides/core/apps.py:
```diff
@@ -19,10 +19,12 @@ def ready(self):
             ChatMessageViewSet,
             ConfigViewSet,
+            CountdownViewSet,
+            MessageViewSet,
```
ProjectorMessageViewSet

normanjaeckel commented on openslides/core/migrations/0008_countdown_message.py:
```diff
@@ -0,0 +1,109 @@
+    Adds projectiondefaults for messages and countdowns.
+    """
+    Projector = apps.get_model('core', 'Projector')
+    ProjectionDefault = apps.get_model('core', 'projectionDefault')
```
`ProjectionDefault = apps.get_model('core', 'ProjectionDefault')`

FinnStutzenstein replied:
Done.

normanjaeckel commented on openslides/core/migrations/0008_countdown_message.py:
```diff
@@ -0,0 +1,109 @@
+        projector=default_projector))
+
+    # Create all new projectiondefaults
+    for projectiondefault in projectiondefaults:
```
Here you can use bulk_create.

normanjaeckel commented on openslides/core/projector.py:
```diff
@@ -15,134 +10,41 @@ class Clock(ProjectorElement):
 
 
-class Message(ProjectorElement):
+class MessageElement(ProjectorElement):
```
`ProjectorMessageElement`

FinnStutzenstein replied:
Done.

normanjaeckel commented on openslides/core/serializers.py:
```diff
@@ -61,3 +68,21 @@ class Meta:
+
+class MessageSerializer(ModelSerializer):
+    """
+    Serializer for core.models.Message objects.
```
core.models.ProjectorMessage

FinnStutzenstein commented on openslides/core/access_permissions.py:
```diff
@@ -64,6 +64,44 @@ def get_serializer_class(self, user=None):
+        """
+        Returns serializer class.
+        """
+        from .serializers import MessageSerializer
```
Done.

FinnStutzenstein commented on openslides/core/access_permissions.py:
```diff
@@ -64,6 +64,44 @@ def get_serializer_class(self, user=None):
 
+class MessageAccessPermissions(BaseAccessPermissions):
+    """
+    Access permissions for Message.
```
Done.

FinnStutzenstein commented on openslides/core/access_permissions.py:
```diff
@@ -64,6 +64,44 @@ def get_serializer_class(self, user=None):
         return ChatMessageSerializer
 
 
+class MessageAccessPermissions(BaseAccessPermissions):
```
Done.

FinnStutzenstein commented on openslides/agenda/models.py:
```diff
@@ -412,8 +412,11 @@ def begin_speech(self):
+            countdown, _ = ProjectorCountdown.objects.get_or_create(pk=1, defaults={
+                'default_time': config['projector_default_countdown'],
+                'countdown_time': config['projector_default_countdown']})
+            countdown.control(action='reset')
```
Done.

FinnStutzenstein commented on openslides/agenda/models.py:
```diff
@@ -422,7 +425,10 @@ def end_speech(self):
         self.save()
         if config['agenda_couple_countdown_and_speakers']:
-            Countdown.control(action='stop')
+            countdown, _ = ProjectorCountdown.objects.get_or_create(pk=1, defaults={
```
Done.

FinnStutzenstein commented:
@normanjaeckel all Done. I've removed the identifier field, renamed ProjectorCountdown back to Countdown and added the Projector prefix for the serializers etc. Also modified the migration and the list of speakers countdown coupling.

emanuelschuetze commented on openslides/core/static/js/core/site.js:
```diff
@@ -908,205 +896,45 @@ angular.module('OpenSlidesApp.core.site', [
-                    stable: true,
-                }]);
-            });
+            var message = {message: gettextCatalog.getString('Message')};
```
Why do you create a default text for a new message? - Now you have to remove this default text first to add custom message text. I prefere to remove this default text and create an empty message.
